File: src/java/magmac/app/lang/node/OperationDeserializer.java
package magmac.app.lang.node;

import magmac.api.Option;
import magmac.app.compile.error.CompileResult;
import magmac.app.compile.node.Node;
import magmac.app.lang.Destructors;
import magmac.app.lang.java.JavaDeserializers;
import magmac.app.lang.java.JavaLang.JavaOperation;

public record OperationDeserializer(Operator operator) implements TypedDeserializer<JavaOperation> {
    @Override
    public Option<CompileResult<JavaOperation>> deserialize(Node node) {
        return Destructors.destructWithType(this.operator().type(), node).map(deserializer -> deserializer.withNode("left", JavaDeserializers::deserializeJavaOrError)
                .withNode("right", JavaDeserializers::deserializeJavaOrError)
                .complete(tuple -> new JavaOperation(tuple.left(), this.operator(), tuple.right())));
    }
}

File: src/web/magmac/app/lang/node/OperationDeserializer.ts
import { Option } from "../../../../magmac/api/Option";
import { CompileResult } from "../../../../magmac/app/compile/error/CompileResult";
import { Node } from "../../../../magmac/app/compile/node/Node";
import { Destructors } from "../../../../magmac/app/lang/Destructors";
import { JavaDeserializers } from "../../../../magmac/app/lang/java/JavaDeserializers";
import { operation } from "../../../../magmac/app/lang/java/JavaLang/operation";
export class OperationDeserializer {
	deserialize(node : Node) : Option<CompileResult<operation>> {return Destructors.destructWithType( this.operator( ).type( ), node).map( 0);;}
}
